src/components/context/TimersContextProvider.tsx

```diff
@@ -0,0 +1,175 @@
+    stopWorkout: () => {},
+    fastForward: () => {},
+});
+
+const TimersContextProvider = ({ children }: { children: React.ReactNode }) => {
+    const [timers, setTimers] = useState<Timer[]>([]);
+    const [running, setRunning] = useState<Timer | Partial<Timer> | null>(null);
+    const intervalRef: MutableRefObject<number | undefined> = useRef();
+    useEffect(() => {
+        if (running === null) {
+            clearInterval(intervalRef.current);
+        }
+    }, [running]);
+
+    useEffect(() => {
+        return () => {
+            clearInterval(intervalRef.current);
+        };
+    }, []);
+
+    return (
+        <TimersContext.Provider
+            value={{
+                timers,
+                running,
+                setRunning,
+                createTimer: ({ type, duration, initialDuration, rounds, initialRounds, restDuration, initialRestDuration }) => {
+                    const newTimer = {
+                        id: `${Date.now()}`,
+                        status: 'stopped',
+                        type,
+                        duration,
+                        initialDuration,
+                        initialRounds,
+                        rounds,
+                        initialRestDuration,
+                        restDuration,
+                    };
+                    setTimers([...timers, newTimer]);
+                    return newTimer;
+                },
+                deleteTimer: id => setTimers(timers.filter(t => t.id !== id)),
+                startWorkout: () => {
+                    if (running === null) {
+                        intervalRef.current = setInterval(() => {
+                            setTimers(prevTimers => {
+                                //We get the index ot the current timer on the array.
+                                const timerIndex = prevTimers.findIndex(timer => timer.status === 'stopped' || timer.status === 'running');
+
+                                //If no timer is left to run we restart the workout.
+                                if (timerIndex === -1) {
+                                    stopWorkout(setTimers, setRunning);
+                                    return prevTimers;
+                                }
+
+                                const updatedTimers = [...prevTimers];
+                                const nextTimer = { ...updatedTimers[timerIndex] };
+
+                                //We increase or decrease the time depending on the type of timer, stopwatch is the only one that goes forward in time.
+                                const timeIncrease = nextTimer.type === 'Stopwatch' ? 1000 : -1000;
+                                nextTimer.duration += timeIncrease;
+
+                                // The logic for the 4 timers is handled here.
+                                if (nextTimer.duration <= 0) {
+                                    if (nextTimer.rounds !== undefined && nextTimer.rounds > 1) {
```

Comment:
would feel a bit cleaner to define the business logic here using the timer type rather than the variables it has defined. That is, `if (type == 'TABATA') do this` etc

Reply:
I had that idea but I try to avoid if possible to define logic based on strings, in general as far as I know a number evaluation is faster than a string evaluation so my logic here was that in general all our timers go forward with the exception of "countdown" and you know what time of timer is each based on:

- restDuration is only used and defined by Tabata timers so whenever a value is greater than 0 it is a Tabata
- rounds are only used by Tabata and XY but XY doesn't define a rest duration so you know it is an XY timer when you have rounds but restTime = 0
- The only exception to the rule is the countdown method which I had to compare with the type to determine the logic.

I can see the benefits of being easier to maintain if the logic was just determined by the type. 😅